katsdpcontroller/product_controller.py

```diff
@@ -907,6 +899,43 @@ async def gain_impl(
         """
         raise NotImplementedError()
 
+    async def gain_all(self, stream_name: str, values: Sequence[str]) -> None:
+        """Set F-engine gains for all inputs."""
+        if self.state not in {ProductState.CAPTURING, ProductState.IDLE}:
+            raise FailReply(f"Cannot set gains in state {self.state}")
+        stream = self._find_stream(stream_name)
+        if not isinstance(stream, product_config.GpucbfAntennaChannelisedVoltageStream):
+            raise FailReply(f"Stream {stream_name!r} is of the wrong type")
+        if len(values) not in {1, stream.n_chans}:
+            raise FailReply(f"Expected 0, 1, or {stream.n_chans} values, received {len(values)}")
+        values = tuple(values)
+        if values != ("default",):
+            try:
+                for v in values:
+                    complex(v)  # Evaluated just for the exception
+            except ValueError as exc:
+                raise FailReply(str(exc))
+        # TODO: remove the type: ignores once SDPSubarrayProduct and SDPSubarrayProductBase merged
+        await self._multi_request(  # type: ignore
+            self.find_nodes(task_type="f", streams=[stream]),  # type: ignore
+            itertools.repeat(("gain-all",) + values)
+        )
+
+    async def capture_start_stop(self, stream_name: str, start: bool) -> None:
+        """Either start or stop transmission on a stream."""
```

Comment:
On the one hand I'm somewhat uncomfortable with a function that starts or stops depending on the bool that gets passed in.

On the other hand, I think I can see why you did this, because there would be a lot of repetition if you had separate functions for each, since they're both basically the same.

As a compromise, I'd suggest putting a `*` in the argument list before the `start` parameter to force the caller to use a named argument. That way when I see
`capture_start_stop("my_stream", start=False)`, then I know it's not starting so it must be stopping.

Alternatively a pattern that I've seen used is to have something like 
```python
def capture_start_stop(action: str):
    if action not in {"start", "stop"}:
        raise ValueError("Expected `start` or `stop`")
    else:
        ...
```

I may just be bike-shedding at this point though. It's difficult to tell.

Reply:
I like the named argument idea. An alternative would be to use an Enum (better than a string since mypy can do the type check statically), but that feels overkill for a single use.